File: agentproof-python/python/aegis_enforce/_engine.py
"""
PolicyEngine — the core evaluation interface.

Wraps the native Rust engine with a Pythonic API. Falls back to a
pure-Python JSON-based evaluator if the native extension isn't built.
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Optional, Union

try:
    from aegis_enforce._agentproof_core import PolicyEngine as _NativeEngine

    _NATIVE_AVAILABLE = True
except ImportError:
    _NATIVE_AVAILABLE = False
    _NativeEngine = None


@dataclass
class PolicyResult:
    """The result of evaluating an event against a policy."""

    verdict: str
    """'allow', 'deny', 'audit', or 'redact'"""

    reason: Optional[str] = None
    """Human-readable explanation for the verdict."""

    triggered_rules: list[int] = field(default_factory=list)
    """IDs of rules that matched the event."""

    actions: list[dict[str, Any]] = field(default_factory=list)
    """Actions to execute (log, notify, escalate, etc.)."""

    violations: list[dict[str, Any]] = field(default_factory=list)
    """Invariant violations detected."""

    constraint_violations: list[dict[str, Any]] = field(default_factory=list)
    """Rate limit / quota violations."""

    eval_time_us: int = 0
    """Evaluation latency in microseconds."""

    @property
    def allowed(self) -> bool:
        """True if the verdict permits the action."""
        return self.verdict in ("allow", "audit")

    @property
    def denied(self) -> bool:
        """True if the verdict blocks the action."""
        return self.verdict == "deny"

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> PolicyResult:
        return cls(
            verdict=d.get("verdict", "allow"),
            reason=d.get("reason"),
            triggered_rules=d.get("triggered_rules", []),
            actions=d.get("actions", []),
            violations=d.get("violations", []),
            constraint_violations=d.get("constraint_violations", []),
            eval_time_us=d.get("eval_time_us", 0),
        )


class PolicyEngine:
    """
    The AgentProof policy engine.

    Loads a compiled .aegisc policy and evaluates agent events against it.
    Uses the native Rust engine when available, falls back to pure Python.

    Examples:
        # From a compiled file
        engine = PolicyEngine.from_file("guard.aegisc")

        # From raw bytes (e.g., loaded from S3, database, etc.)
        engine = PolicyEngine.from_bytes(policy_bytes)

        # Evaluate an event
        result = engine.evaluate("tool_call", {
            "tool": "http_request",
            "url": "https://api.external.com/data",
        })

        if result.denied:
            raise PermissionError(result.reason)
    """

    def __init__(self, native_engine=None, policy_name: str = "unknown"):
        self._engine = native_engine
        self._policy_name = policy_name
        self._event_count = 0

    @classmethod
    def from_file(cls, path: Union[str, Path]) -> PolicyEngine:
        """Load a policy from a compiled .aegisc file."""
        path = Path(path)
        if not path.exists():
            raise FileNotFoundError(f"Policy file not found: {path}")

        if _NATIVE_AVAILABLE:
            engine = _NativeEngine.from_file(str(path))
            return cls(native_engine=engine, policy_name=engine.policy_name)
        else:
            # Pure Python fallback: read and parse the JSON payload
            return cls._from_file_pure_python(path)

    @classmethod
    def from_bytes(cls, data: bytes) -> PolicyEngine:
        """Load a policy from raw .aegisc bytes."""
        if _NATIVE_AVAILABLE:
            engine = _NativeEngine.from_bytes(data)
            return cls(native_engine=engine, policy_name=engine.policy_name)
        else:
            return cls._from_bytes_pure_python(data)

    @classmethod
    def from_json(cls, json_str: str) -> PolicyEngine:
        """Load a policy from a JSON string (debug format)."""
        if _NATIVE_AVAILABLE:
            engine = _NativeEngine.from_json(json_str)
            return cls(native_engine=engine, policy_name=engine.policy_name)
        else:
            policy = json.loads(json_str)
            return cls(policy_name=policy.get("name", "unknown"))

    def evaluate(
        self,
        event_type: str,
        fields: Optional[dict[str, Any]] = None,
    ) -> PolicyResult:
        """
        Evaluate an event against the loaded policy.

        Args:
            event_type: The event type (e.g., "tool_call", "data_access")
            fields: Event-specific fields as a dict

        Returns:
            A PolicyResult with the verdict and metadata.

        Raises:
            RuntimeError: If the engine is not properly initialized.
        """
        self._event_count += 1

        if self._engine is not None:
            raw = self._engine.evaluate(event_type, fields)
            return PolicyResult.from_dict(raw)
        else:
            # Pure Python fallback: allow everything (no enforcement)
            return PolicyResult(
                verdict="allow",
                reason="No native engine available (pure Python fallback)",
                eval_time_us=0,
            )

    def set_context(self, key: str, value: Any) -> None:
        """Set a persistent context value (survives across events)."""
        if self._engine is not None:
            self._engine.set_context(key, value)

    def set_config(self, key: str, value: Any) -> None:
        """Set a policy configuration value."""
        if self._engine is not None:
            self._engine.set_config(key, value)

    @property
    def policy_name(self) -> str:
        """The name of the loaded policy."""
        return self._policy_name

    @property
    def event_count(self) -> int:
        """Total number of events evaluated."""
        if self._engine is not None:
            return self._engine.event_count
        return self._event_count

    def reset(self) -> None:
        """Reset engine state (state machines, counters)."""
        if self._engine is not None:
            self._engine.reset()
        self._event_count = 0

    def status(self) -> str:
        """Get a human-readable status summary."""
        if self._engine is not None:
            return self._engine.status()
        return f"PolicyEngine(policy='{self._policy_name}', native=False)"

    def __repr__(self) -> str:
        native = "native" if self._engine is not None else "python"
        return f"<PolicyEngine policy='{self._policy_name}' engine={native}>"

    # ── Pure Python fallback (reads .aegisc without native code) ──────

    @classmethod
    def _from_file_pure_python(cls, path: Path) -> PolicyEngine:
        """Read .aegisc file format in pure Python."""
        with open(path, "rb") as f:
            magic = f.read(4)
            if magic != b"\xae\x91\x5c\x01":
                raise ValueError(f"Not a valid .aegisc file: {path}")
            _version = int.from_bytes(f.read(2), "little")
            _flags = int.from_bytes(f.read(2), "little")
            payload_len = int.from_bytes(f.read(4), "little")
            payload = f.read(payload_len)

        policy = json.loads(payload)
        return cls(policy_name=policy.get("name", "unknown"))

    @classmethod
    def _from_bytes_pure_python(cls, data: bytes) -> PolicyEngine:
        """Parse .aegisc bytes in pure Python."""
        if data[:4] != b"\xae\x91\x5c\x01":
            raise ValueError("Not valid .aegisc data")
        payload_len = int.from_bytes(data[8:12], "little")
        payload = data[12 : 12 + payload_len]
        policy = json.loads(payload)
        return cls(policy_name=policy.get("name", "unknown"))

File: automaguard-rs/src/engine.rs
//! Synchronous policy engine — the primary entry point for the Rust SDK.

use std::collections::HashMap;
use std::path::Path;

use smol_str::SmolStr;

use aegis_compiler::ast::Verdict;
use aegis_compiler::bytecode;
use aegis_runtime::engine::{PolicyEngine as RuntimeEngine, PolicyResult as RuntimeResult};
use aegis_runtime::event::{Event, Value};

use crate::error::Error;

// ── PolicyResult ──────────────────────────────────────────────────────────────

/// The result of evaluating a single event against a loaded policy.
///
/// Wraps [`aegis_runtime::engine::PolicyResult`] with convenience methods.
/// Obtain one by calling [`PolicyEngine::evaluate`].
#[derive(Debug, Clone)]
pub struct PolicyResult(pub(crate) RuntimeResult);

impl PolicyResult {
    /// Returns `true` if the verdict is `Deny`.
    pub fn is_denied(&self) -> bool {
        self.0.verdict == Verdict::Deny
    }

    /// Returns `true` if the verdict is `Allow`.
    pub fn is_allowed(&self) -> bool {
        self.0.verdict == Verdict::Allow
    }

    /// Returns `true` if the verdict is `Audit` (log and allow).
    pub fn is_audit(&self) -> bool {
        self.0.verdict == Verdict::Audit
    }

    /// Returns `true` if the verdict is `Redact` (sanitise and allow).
    pub fn is_redact(&self) -> bool {
        self.0.verdict == Verdict::Redact
    }

    /// The human-readable reason accompanying the verdict, if any.
    pub fn reason(&self) -> Option<&str> {
        self.0.reason.as_deref()
    }

    /// The IDs of rules that fired for this event.
    pub fn triggered_rules(&self) -> &[u32] {
        &self.0.triggered_rules
    }

    /// Temporal invariant violations detected during this evaluation.
    pub fn violations(&self) -> &[aegis_runtime::engine::Violation] {
        &self.0.violations
    }

    /// Rate-limit or quota constraint violations detected during this evaluation.
    pub fn constraint_violations(&self) -> &[aegis_runtime::engine::ConstraintViolation] {
        &self.0.constraint_violations
    }

    /// How long the evaluation took, in microseconds.
    pub fn latency_us(&self) -> u64 {
        self.0.eval_time_us
    }

    /// The raw [`Verdict`] value.
    pub fn verdict(&self) -> Verdict {
        self.0.verdict
    }

    /// Borrow the underlying runtime result.
    pub fn inner(&self) -> &RuntimeResult {
        &self.0
    }
}

impl std::fmt::Display for PolicyResult {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}", self.0.verdict)?;
        if let Some(reason) = self.reason() {
            write!(f, ": {reason}")?;
        }
        Ok(())
    }
}

// ── PolicyEngine ──────────────────────────────────────────────────────────────

/// Synchronous AutomaGuard policy engine.
///
/// Loads a compiled `.aegisc` policy once at construction and evaluates agent
/// events against it at sub-microsecond latency.
///
/// # Example
///
/// ```ignore
/// use automaguard::{PolicyEngine, EnforcementError};
/// use std::collections::HashMap;
///
/// let mut engine = PolicyEngine::from_file("guard.aegisc")?;
///
/// let mut fields = HashMap::new();
/// fields.insert("tool_name".into(), "send_email".into());
///
/// let result = engine.evaluate("tool_call", fields)?;
/// if result.is_denied() {
///     return Err(EnforcementError::new(result).into());
/// }
/// ```
pub struct PolicyEngine {
    inner: RuntimeEngine,
}

impl std::fmt::Debug for PolicyEngine {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("PolicyEngine")
            .field("policy", &self.inner.policy_name())
            .field("events_processed", &self.inner.event_count())
            .finish()
    }
}

impl PolicyEngine {
    /// Load a compiled policy from a `.aegisc` file on disk.
    pub fn from_file(path: impl AsRef<Path>) -> Result<Self, Error> {
        let policy = bytecode::read_file(path.as_ref())?;
        Ok(Self {
            inner: RuntimeEngine::new(policy),
        })
    }

    /// Load a compiled policy from an in-memory `.aegisc` byte slice.
    pub fn from_bytes(bytes: &[u8]) -> Result<Self, Error> {
        let policy = bytecode::from_bytecode(bytes)?;
        Ok(Self {
            inner: RuntimeEngine::new(policy),
        })
    }

    /// Evaluate a single agent event against the loaded policy.
    ///
    /// - `event_type` — the event category (e.g. `"tool_call"`, `"data_access"`).
    /// - `fields`     — key-value map of event fields. Keys accept anything that
    ///   converts to [`SmolStr`]; values are [`Value`] instances.
    ///
    /// Returns a [`PolicyResult`] containing the verdict, triggered rules, and
    /// any invariant or rate-limit violations. The `?` operator is for
    /// infrastructure errors only (e.g. future async transport failures); in v1
    /// the synchronous path never returns `Err`.
    pub fn evaluate(
        &mut self,
        event_type: &str,
        fields: HashMap<SmolStr, Value>,
    ) -> Result<PolicyResult, Error> {
        let event = Event::new(event_type).with_fields(fields);
        Ok(PolicyResult(self.inner.evaluate(&event)))
    }

    /// Convenience builder for evaluating events with a fluent field API.
    ///
    /// ```ignore
    /// let result = engine.event("tool_call")
    ///     .field("tool_name", "search")
    ///     .field("query", "customer records")
    ///     .evaluate()?;
    /// ```
    pub fn event<'e>(&'e mut self, event_type: &str) -> EventBuilder<'e> {
        EventBuilder {
            engine: self,
            event_type: event_type.to_owned(),
            fields: HashMap::new(),
        }
    }

    /// Set a persistent context value that survives across evaluations.
    ///
    /// Context values are accessible from policy expressions via the `context`
    /// root (e.g. `context.user_role`).
    pub fn set_context(&mut self, key: impl Into<SmolStr>, value: Value) {
        self.inner.set_context(key, value);
    }

    /// The name of the loaded policy.
    pub fn policy_name(&self) -> &str {
        self.inner.policy_name()
    }

    /// Total number of events evaluated by this engine instance.
    pub fn event_count(&self) -> u64 {
        self.inner.event_count()
    }

    /// Reset all state machines and rate limiters to their initial state.
    ///
    /// Useful in tests or when starting a new agent session on a reused engine.
    pub fn reset(&mut self) {
        self.inner.reset();
    }
}

// ── EventBuilder ──────────────────────────────────────────────────────────────

/// Fluent builder for constructing and evaluating an event in one expression.
///
/// Obtain via [`PolicyEngine::event`].
pub struct EventBuilder<'e> {
    engine: &'e mut PolicyEngine,
    event_type: String,
    fields: HashMap<SmolStr, Value>,
}

impl<'e> EventBuilder<'e> {
    /// Add a field to the event.
    ///
    /// `key` accepts anything that converts to [`SmolStr`] (e.g. `&str`).
    /// `value` accepts anything that converts to [`Value`].
    pub fn field(mut self, key: impl Into<SmolStr>, value: impl Into<Value>) -> Self {
        self.fields.insert(key.into(), value.into());
        self
    }

    /// Evaluate the event and return the policy result.
    pub fn evaluate(self) -> Result<PolicyResult, Error> {
        self.engine.evaluate(&self.event_type, self.fields)
    }
}

// Value From impls live in aegis-runtime/src/event.rs (where Value is defined).
